Prevent cmd.exe shell windows popping up in the background when calling subprocess

cjrh opened the pull request:
I had many fights with line-endings!  It looks like some of the patch and diff tools between mercurial and git don't respect the various autocrlf settings in dot files.

ccordoba12 commented:
Hi @cjrh glad to see you on this side!! Could you change the PR title to reflect what you're actually doing here?

cjrh commented:
Done.

ccordoba12 commented:
Ok, two things:

1. This needs a rebase
2. Is this ready to be merged?

cjrh commented:
Right now I'm merging the latest master into this PR.  If all goes well we should be able to merge. I'll report back when I'm done.

cjrh commented on spyderlib/widgets/externalshell/sitecustomize.py:
```diff
@@ -831,10 +833,10 @@ def evalsc(command):
         if os.name == 'nt':
-            Popen('dir', shell=True, stdin=PIPE)
+            programs.run_shell_command('dir', stdin=PIPE)
             _print('\n')
```
The `_print()` can be moved outside the `if` block.

ccordoba12 commented:
@blink1073, could you help me to review this one?

cjrh commented:
@ccordoba12 It should be ok for review and testing now. Picked up one or two regressions where names changed in master, but merge was successful (scary!). I did superficial testing in Windows 8.1 x64, on both Python 2.7.9 and Python 3.4.1 (Anaconda) on PyQt, as well as my (main) OS X Yosemite, Python 2.7.9 (again Anaconda).

The nature of the changes is somewhat complex and prone to subtle bugs.  So bear that in mind when you test.

In general, if this PR is merged, it would mean that all system calls in the future should go either through `programs.run_shell_command()` *or* `programs.run_program()`. I tried *really hard* to find a way to use one command to handle both case, but I could not get away from the mess that is `Popen`. Separating shell commands and subprocesses makes things much easier to handle and understand what is going on.

blink1073 commented on spyderlib/utils/codeanalysis.py:
```diff
@@ -144,8 +144,10 @@ def check(args, source_code, filename=None, options=None):
-    output = Popen(args, stdout=PIPE, stderr=PIPE
-                   ).communicate()[0].strip().decode().splitlines()
+    cmd = args[0]
+    cmdargs = [] if len(args) < 2 else args[1:]
```
This can just be written as `args[1:]`.

blink1073 commented:
I'd argue it would be cleaner to use `kwargs.setdefault('stderr', subprocess.PIPE)` and `kwargs.setdefault('stderr', subprocess.PIPE)` rather than forcing all the other users of `programs.run_program` to import from `subprocess` and specify them every time.

cjrh commented:
f.y.i. I added @ccordoba12 and @blink1073 as collaborators to my Spyder fork (where this PR comes from).  I have no problem with you guys tweaking things in my fork 👍

blink1073 commented:
@cjrh, you didn't rebase, you just merged from master  :smile:

Assuming you have this repo set as `upstream`:

```bash
$ git fetch upstream master
$ git rebase -i upstream master
```

cjrh commented:
@blink1073 Ok, if that's the preferred way for the project I can do it. I have some personal horror stories involving rebase.  Also [this](http://geekblog.oneandoneis2.org/index.php/2013/04/30/please-stay-away-from-rebase).

Nodd commented on spyderlib/utils/programs.py:
```diff
@@ -55,13 +55,79 @@ def find_program(basename):
+    return subprocess.Popen(cmdstr, **kwargs)
+
+
+def run_program(program, args=[], **subprocess_kwargs):
```
Using `args=None` here and checking `if args:` later may avoid possible bugs in the future, even if right now `args` is not modified so it works fine.

cjrh replied:
@Nodd Will change thanks. Originally, the reason I chose `args=[]` rather than `args=None` was because there were no doctoring hints as to the required type of args, and having the default be an empty list instead of None will give the caller a hint (e.g. if they use jedi for completion etc.).  Since that first draft I have added sphinx types in the docstring, so editors will show the docstring types and we don't need to hint it with the default type.  I will change it to `None`.

Nodd commented on spyderlib/utils/programs.py:
```diff
@@ -55,13 +55,79 @@ def find_program(basename):
+    
+    :str cmdstr: The string run as a shell command.
+    :subprocess_kwargs: These will be passed to subprocess.Popen."""
+    kwargs = subprocess_kwargs
```
Why not keeping `subprocess_kwargs` in the whole function ? (same for `run_program`)

Nodd commented:
Having to different functions is indeed cleaner, but why not declaring them as "private" and use one intermediate function that would call the correct subfunction depending on the arguments ? This way the implementation is clean and separated, and using it is easy because it's only one function.

Another option for `PIPE` is to add it to the `programs` namespace so that it is possible to use `programs.PIPE`. This is just an alternative, setting PIPE as default makes sense too.

cjrh commented:
@Nodd Thanks for your line suggestions, I made both changes. (`args=[]` was intended to hint to callers that the type was `list` not `str`, but that was before I added sphinx types in the docstring, making type-hinting-by-default-type redundant).

With respect to "Why not one intermediate function?": The problem is that the confusion sits on the side of the caller, not the implementation. I'm not smart enough to figure out an easy-to-understand API for handling both shell cmd and programs in one function interface that improves at all on the `Popen` methods, and I *hate* the subprocess Popen API.  In building this PR, I tried a bunch of different ways to do it better, but I failed repeatedly.

Nodd commented:
Fine, having two functions also forces the dev to think about what he's really doing so it has advantages too !

blink1073 commented:
@ccordoba12, I tried this on Windows (open a new terminal, run in a new dedicated terminal), and everything works fine.

ccordoba12 commented:
@blink1073, the problem is when you open our Preferences, which calls `subprocess` in the background to run some tests and change some options on the fly.

blink1073 commented:
@ccordoba12, I tried what you suggested on Py2.7 and 3.4, but I don't get the popup windows even on the master branch.  This branch at least doesn't break anything  :smile:

cjrh commented:
@blink1073 On my very fast mac, even when running Windows in a VM, the cmd.exe windows flash almost too fast to see.  Are you on a fast machine?

blink1073 commented:
That could very well be it, if only I still had my netbook with Windows 7.

goanpeca commented:
I have seen this in my work laptop from time to time, I will check today any other way to check it??? 

Why not add a (temporal)  `time.sleep(5)` when it is called to ensure it is indeed not there? 

That is how I tested the restart PR (to make sure there were no windows popping up)

goanpeca commented:
I just checked... I do not see any cmd prompt flashing...

goanpeca commented:
@ccordoba12 @cjrh merging?

cjrh commented:
I've updated to latest master and...no more popup windows are happening.  Let me first see what is going on.

cjrh commented:
@ccordoba12 Do you still see the CMD shell windows popping up on master, e.g. when clicking "Help > About" (fetches the git hash) or with "Tools > Preferences" (launches multiple shell commands)?  All of a sudden, these consoles no longer appear, even if I roll back to before I made this PR in the firs place. I feel like I'm losing my mind o.O

cjrh commented:
@goanpeca If you get a chance, could you please launch Spyder bootstrap like this:
```batch
start pythonw bootstrap.py --gui=pyqt -- --new-instance
```
Then try "Help > About" and "Tools > Preferences".

This is how my own desktop shortcuts are launching Spyder.  It turns out there is a big difference between `python.exe` and `pythonw.exe` when `subprocess` calls are involved.

cjrh commented:
Ok, here is the situation observable from my side:
- The Spyder launcher shortcut in Anaconda Python (latest), in the Windows Start menu, uses
```batch
F:\Anaconda\pythonw.exe "F:\Anaconda\Scripts/spyder-script.py"
```
- The reason this is done (instead of using `python.exe`) is so that there is no cmd shell window attached to Spyder, floating in the background.
- However, this produces the issue for which this PR was made, i.e. calls via `subprocess` make cmd shell windows pop up each time. *Note: this PR does fix this issue.*
- My own startup shortcuts for Spyder use `pythonw`, which I why I incorrectly thought this was a systematic issue
- It turns out that if you launch with `python.exe bootstrap`, then `subprocess` does not create flashing background cmd shell windows.

Yay! So I'm not losing my mind after all 👍

blink1073 commented:
@cjrh, I can *finally* reproduce.  When I run Tools->Preferences on Master, I get several windows popping up, and your PR clears that up.  However, for Help->About, I get a single popup window with and without the PR.

cjrh commented:
I made a tester app that you can use to see the effect of starting a PyQt app either via `python.exe` or via `pythonw.exe`.  The app has four buttons that use `subprocess` to call `git --help` in various ways, and retrieve the output.  

So try:

0. `C:\> python main.py`
0. `C:\> pythonw main.py`

You should see that `python` always requires a console window attached to the app (undesirable), while `pythonw` makes cmd shell windows flash for the *non-shell* version of `Popen`.  I included `check_output()` calls, just to show that you must supply `stdin` and possibly `stderr`, but must not provide `stdout`.  It actually won't work if you don't supply `stdin` to the call.

So this PR makes it so that a PyQt app that is launched with `pythonw` will also not show flashing windows.

---

File **testpopen.py**:

```python
import subprocess
from subprocess import PIPE

streams = dict(stdin=PIPE, stdout=PIPE, stderr=PIPE)

def popen(shell=False):
    arg = ['git', '--help']
    if shell:
        arg = ' '.join(arg)
    cm = subprocess.Popen(
            arg,
            shell=shell,
            **streams
            )
    cmc = cm.communicate()
    return cmc[0]


def progchk():
    output = subprocess.check_output(
            ['git', '--help'],
# https://docs.python.org/2/library/subprocess.html#subprocess.check_output            
            **{k:v for k,v in streams.iteritems() if not k=='stdout'}
            )
    return output

def shellchk():
    output = subprocess.check_output(
            'git --help',
            shell=True,
# https://docs.python.org/2/library/subprocess.html#subprocess.check_output            
            **{k:v for k,v in streams.iteritems() if not k=='stdout'}
            )
    return output

if __name__=='__main__':
    print popen(shell=True)
    print popen(shell=False)

    print shellchk()
    print progchk()
```

File **main.py**:
```python
from PyQt4 import QtCore, QtGui
import testpopen
import sys


class MainForm(QtGui.QMainWindow):
    def __init__(self, *args, **kwargs):
        super(MainForm, self).__init__(*args, **kwargs)

        self.b1 = QtGui.QPushButton('prog', self)
        self.b1.clicked.connect(self.runprog)

        self.b2 = QtGui.QPushButton('shell', self)
        self.b2.clicked.connect(self.runshell)

        self.b3 = QtGui.QPushButton('prog check_output', self)
        self.b3.clicked.connect(self.runprogchk)

        self.b4 = QtGui.QPushButton('shell check_output', self)
        self.b4.clicked.connect(self.runshellchk)

        self.edit = QtGui.QTextEdit('', self)

        self.qw = QtGui.QWidget()
        self.setCentralWidget(self.qw)

        self.vbox = QtGui.QVBoxLayout()
        self.vbox.addWidget(self.b1)
        self.vbox.addWidget(self.b2)
        self.vbox.addWidget(self.b3)
        self.vbox.addWidget(self.b4)
        self.vbox.addWidget(self.edit)
        self.qw.setLayout(self.vbox)

    def runprog(self):
        self.edit.setText('')
        result = testpopen.popen()
        self.edit.setText('PROG:\n\n'+result)

    def runshell(self):
        self.edit.setText('')
        result = testpopen.popen(shell=True)
        self.edit.setText('SHELL:\n\n'+result)

    def runprogchk(self):
        self.edit.setText('')
        result = testpopen.progchk()
        self.edit.setText('PROGCHK:\n\n'+result)

    def runshellchk(self):
        self.edit.setText('')
        result = testpopen.shellchk()
        self.edit.setText('SHELLCHK:\n\n'+result)

if __name__ == '__main__':
    app = QtGui.QApplication(sys.argv)
    form = MainForm()
    form.show()
    sys.exit(app.exec_())
```

cjrh commented:
@blink1073 It'll be some minor fix. I know that I had this working, but the last merge was a bit messy.  I will sort it out.

cjrh commented:
I'm thinking the Python docs for subprocess really need some improvement. I have had the docs open in front of me for many, many hours, but I still have to resort to poke testing to see what happens for different combinations of options 😢

blink1073 commented:
Agreed, sorry. :(

cjrh commented:
@blink1073 Ok I fixed the "About" window (I missed the git section in the changeover from mercurial).

I think I should probably hide all the `subprocess.PIPE` stuff under `kwargs.setdefault()` inside `run_program()` as per your earlier suggestion.  Are there instances where default values of `PIPE` for the io streams would not make sense?

blink1073 commented:
I think it always makes sense for a managed process.

cjrh commented:
Changed to setting the io streams to use `PIPE` by default, unless specified in the args to `run_program()` or `run_shell_command()`. I also refactored the caller code to take this into account, cleaning it up nicely.

I made separate commits for the refactoring to each file, just in case we need to revert one or two of them for some PIPE-related issue that I wasn't expecting.

Nodd commented on spyderlib/utils/programs.py:
```diff
@@ -55,13 +59,92 @@ def find_program(basename):
+    if os.name == 'nt':
+        CONSOLE_CREATION_FLAGS = 0  # Default value
+        # Add more here..
+        CREATE_NO_WINDOW = 0x08000000
```
Where does this magic value come from ?

goanpeca replied:
I believe it comes from [here](https://msdn.microsoft.com/en-us/library/windows/desktop/ms684863%28v=vs.85%29.aspx)

goanpeca replied:
But yes, there should be some link or extra information...

Nodd replied:
Right, a comment with a link to the windows API would be better then.

Nodd commented on spyderplugins/widgets/pylintgui.py:
```diff
@@ -52,10 +51,14 @@ def get_pylint_version():
-                               stdout=subprocess.PIPE, stderr=subprocess.PIPE,
-                               cwd=osp.dirname(PYLINT_PATH),
-                               shell=True if os.name == 'nt' else False)
+    kwargs=dict(cwd=osp.dirname(PYLINT_PATH))
```
missing spaces around the first `=`

Nodd commented on spyderplugins/widgets/pylintgui.py:
```diff
@@ -52,10 +51,14 @@ def get_pylint_version():
+        cmd = PYLINT + ' '.join(args)
+        process = programs.run_shell_command(cmd, **kwargs)
+    else:
+        process = programs.run_program(cmd, args, **kwargs)
```
It would be clearer to define cmd and args in each branch. `cmd` is redefined in the first branch, and the join seems pointless since args is defined just before as a list with one value.

Also I think it lacks a space between `PYLINT` and `' '.join(args)`.

cjrh commented:
Good catches, @Nodd  thank you.  Changes made.

cjrh commented:
Note to self that if (when?) #2410 is merged, I will have to update the code in `vcs.py` here again.

goanpeca commented:
@crjh, is this goign to be updated, rebase?

cjrh commented:
Hi @goanpeca 

Well spotted, I need to update.  I'm incredibly busy at the moment.    I've set a reminder for myself to get it done before the end of 2015. If it still isn't ready by January please poke me with a stick :)

goanpeca commented:
I will :+1:

cjrh commented:
@goanpeca I went through the code and fixed up the parts where master had changed.  I checked that the changes work on Windows (no flashing cmd windows when running from bootstrap), and on mac the behaviour is unchanged.

For some reason, github UI shows me changing README.md and a few other files, but these changes are from master, not me.   My local tools (Sourcetree, Pycharm) do not show these when comparing master and this branch.

Anyhow, what is the easiest way to rebase onto master?  I usually follow merging rather than rebasing and I don't want to screw it up.  When I try a rebase from my GUI tools, it seems to want me to fix merge conflicts again :/

cjrh commented:
@goanpeca The interactive rebase was too complicated (too much history and too many file changes) so  I decided to just squash everything.  I hope that's ok.

blink1073 commented:
I've done the same a few times :)

goanpeca commented:
Hey @cjrh, thanks for :X:-mas update, sure, I am just very used to rebasing now and is wht I usually do, but it is true that it can easily get unwieldy if many files have changed.

So either way is fine, merge, squash and rebase... whatever gets the work done :-)

Is it ready to go?

----

@ccordoba12 any comments here?

cjrh commented:
I think it is, but I would other people to give a sanity check.  After this is merged, `subprocess.Popen` should no longer be used anywhere, but rather `programs.run_shell_command` and `programs.run_program`.  This will ensure the correct flags are set for each platform.

goanpeca commented:
@cjrh great, since this will be something devs will need to be aware of, could you please update https://github.com/spyder-ide/spyder/wiki/Dev%3A-Cookbook

With some examples for both cases and a small explanation?

This will have to go into the developer guide (a work still in progress) eventually

cjrh commented:
@goanpeca Done, see

https://github.com/spyder-ide/spyder/wiki/Dev:-Cookbook

I didn't add examples yet, but there are examples in the source, i.e. the changes in this PR.

cjrh commented:
@goanpeca ok fine I added examples.  I agree it's better to have simple examples right there in the docs.

goanpeca commented:
@cjrh thanks a lot. it looks really good and will be super useful.


:+1:

goanpeca commented:
@ccordoba12 can we merge this?

ccordoba12 commented:
It seems @cjrh gave a lot of thought to this one, and tests are passing, so merging :-)
